drivers/gpio/gpio_xlnx_axi.c

```diff
@@ -229,6 +222,8 @@ static int gpio_xlnx_axi_init(const struct device *dev)
 {
 	struct gpio_xlnx_axi_data *data = dev->data;
 
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
```

Comment:
My understanding is that if dual-channel GPIO is enabled, then the driver will create two devices that will be initialized separately. Will having both devices calling DEVICE_MMIO_MAP cause problems? The second channel will reside in the first's virtual address mapping, but will attempt to call DEVICE_MMIO_MAP with a different size and starting address.

Reply:
I am not sure mapping the same region twice is allowed.

Reply:
If that is the case, would something like this be allowed:

```
#ifdef CONFIG_MMU
static int gpio_xlnx_axi_mmio_init(const struct device *dev)
{
	const struct gpio_xlnx_axi_config *config = dev->config;
	struct gpio_xlnx_axi_data *data = dev->data;

	device_map(&data->mmio, config->base, config->size, K_MEM_CACHE_NONE);

	gpio_xlnx_axi_write_data(dev, data->dout);
	gpio_xlnx_axi_write_tri(dev, data->tri);

	return 0;
}
#endif

static int gpio_xlnx_axi_init(const struct device *dev)
{
	struct gpio_xlnx_axi_data *data = dev->data;

#ifdef CONFIG_MMU
	const device_handle_t *handles;
	size_t handle_count;
	const struct device *parent_device;
	struct gpio_xlnx_axi_data *parent_data;

	handles = device_required_handles_get(dev, &handle_count);

	parent_device = device_from_handle(handles[0]);
	parent_data = parent_device->data;
	data->mmio = parent_data->mmio + GPIO2_DATA_OFFSET;
#endif

	gpio_xlnx_axi_write_data(dev, data->dout);
	gpio_xlnx_axi_write_tri(dev, data->tri);

	return 0;
}
```

The gpio2 nodes would be moved down to `PRE_KERNEL_2` priority as they currently are initialized before their parent node. If `CONFIG_MMU` is defined then the parent gpio node will initialize using `gpio_xlnx_axi_mmio_init()` and will create the memory-mapped region. The gpio2 nodes would initialize using `gpio_xlnx_axi_init()` and would look for their parent to get the base address to write to. 

If `CONFIG_MMU` is not defined then both the gpio and gpio2 nodes will initialize with gpio_xlnx_axi_init.